### PR TITLE
feat: defines internal utils for OAuth client Auth and error handling

### PR DESCRIPTION
**Pull request by bojeil-google**
This defines internal utilities to handle:
- OAuth client authentication
- Standard OAuth error response parsing into OAuthError exception.

### CODE REVIEW

**Comment by busunkim96** [comment]
@bojeil-google Base branch has been adjusted to `byoid`.

**Comment by busunkim96** [comment]
@bojeil-google It may be helpful to join the org `googleapis` since you'll be working on a few PRs in this repo (and in other auth libs?) Kokoro CI won't run automatically for folks outside of the org because of security concerns.

File: google/oauth2/utils.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+
+# OAuth client authentication based on
+# https://tools.ietf.org/html/rfc6749#section-2.3.
+ClientAuthType = enum(BASIC="basic", REQUEST_BODY="request-body")
+
+
+class ClientAuthentication(object):
+    """Defines the client authentication credentials for basic and request-body
+    types based on Based on https://tools.ietf.org/html/rfc6749#section-2.3.1.
```
Comment (busunkim96) [inline google/oauth2/utils.py:71]:
```suggestion
    types based on https://tools.ietf.org/html/rfc6749#section-2.3.1.
```

Comment (bojeil-google) [reply google/oauth2/utils.py]: Done

File: google/oauth2/utils.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+import abc
+import base64
+import json
+
+import six
+
+from google.auth import exceptions
+
+
+def enum(**named_values):
```
[the remark this answers is not in the document]
Comment (bojeil-google) [reply google/oauth2/utils.py]: Ok. I tried it out with the backport `enum34` package . It seems to work locally.

**Review by busunkim96 — APPROVED** [review]

**Pull request merged**